airbyte-integrations/connectors/destination-snowflake-cortex/destination_snowflake_cortex/cortex_processor.py

```diff
@@ -0,0 +1,406 @@
+            MULTI_STATEMENT_COUNT = 0
+            """
+        )
+
+    def _emulated_merge_temp_table_to_final_table(
+        self,
+        stream_name: str,
+        temp_table_name: str,
+        final_table_name: str,
+    ) -> None:
+        """Emulate the merge operation using a series of SQL commands.
+
+        This method varies from the SnowflakeSqlProcessor implementation in that multiple rows will exist for each
+        primary key. And we need to remove all rows (chunks) for a given primary key before inserting new ones.
+
+        So instead of using UPDATE and then INSERT, we will DELETE all rows for included primary keys and then call
+        the append implementation to insert new rows.
+        """
+        columns_list: list[str] = list(
+            self._get_sql_column_definitions(stream_name=stream_name).keys()
+        )
+
+        delete_statement = dedent(
+            f"""
+            DELETE FROM {final_table_name}
+            WHERE {DOCUMENT_ID_COLUMN} IN (
+                SELECT {DOCUMENT_ID_COLUMN}
+                FROM {temp_table_name}
+            );
+            """
+        )
+        append_statement = dedent(
+            f"""
+            INSERT INTO {final_table_name}
+                ({", ".join(columns_list)})
+            SELECT {", ".join(columns_list)}
+            FROM {temp_table_name};
+            """
+        )
+
+        with self.get_sql_connection() as conn:
+            # This is a transactional operation to avoid outages, in case
+            # a user queries the data during the operation.
+            conn.execute(delete_statement)
+            conn.execute(append_statement)
+
+    def process_record_message(
+        self,
+        record_msg: AirbyteRecordMessage,
+        stream_schema: dict,
+    ) -> None:
+        """Write a record to the cache.
+
+        We override the SQLProcessor implementation in order to handle chunking, embedding, etc.
+
+        This method is called for each record message, before the record is written to local file.
+        """
+        document_chunks, id_to_delete = self.splitter.process(record_msg)
+
+        # TODO: Decide if we need to incorporate this into the final implementation:
+        _ = id_to_delete
+
+        for chunk in document_chunks:
+            new_data: dict[str, Any] = {
+                DOCUMENT_ID_COLUMN: self._create_document_id(record_msg),
+                CHUNK_ID_COLUMN: str(uuid.uuid4().int),
+                METADATA_COLUMN: chunk.metadata,
+                DOCUMENT_CONTENT_COLUMN: chunk.page_content,
+                EMBEDDING_COLUMN: chunk.embedding,
+            }
+
+        self.file_writer.process_record_message(
+            record_msg=AirbyteRecordMessage(
+                namespace=record_msg.namespace,
+                stream=record_msg.stream,
+                data=new_data,
+                emitted_at=record_msg.emitted_at,
+            ),
+            stream_schema=stream_schema,
+        )
```

Comment:
Rather than write the raw source record to JSONL, we write the chunked and embedded record to JSONL, and all records will later be loaded in bulk from those files into the Snowflake destination table.